src/main/java/org/apache/commons/imaging/formats/tiff/datareaders/DataReaderStrips.java

```diff
@@ -291,25 +318,50 @@ public BufferedImage readImageData(final Rectangle subImage)
-            final long bytesPerStrip = rowsInThisStrip * bytesPerRow;
-            final long pixelsPerStrip = rowsInThisStrip * width;
-
-            final byte[] compressed = imageData.getImageData(strip).getData();
-
-            final byte[] decompressed = decompress(compressed, compression,
-                    (int) bytesPerStrip, width, (int) rowsInThisStrip);
-
-            interpretStrip(
-                    workingBuilder,
-                    decompressed,
-                    (int) pixelsPerStrip,
-                    yLimit);
+        if (planarConfiguration != 2) {
+            for (int strip = strip0; strip <= strip1; strip++) {
+                final long rowsPerStripLong = 0xFFFFffffL & rowsPerStrip;
+                final long rowsRemaining = height - (strip * rowsPerStripLong);
+                final long rowsInThisStrip = Math.min(rowsRemaining, rowsPerStripLong);
+                final long bytesPerRow = (bitsPerPixel * width + 7) / 8;
+                final long bytesPerStrip = rowsInThisStrip * bytesPerRow;
+                final long pixelsPerStrip = rowsInThisStrip * width;
+
+                final byte[] compressed = imageData.getImageData(strip).getData();
+
+                final byte[] decompressed = decompress(compressed, compression,
+                  (int) bytesPerStrip, width, (int) rowsInThisStrip);
+
+                interpretStrip(
+                  workingBuilder,
+                  decompressed,
+                  (int) pixelsPerStrip,
+                  yLimit);
+            }
+        } else {
+            int nStripsInPlane = imageData.getImageDataLength() / 3;
+            for (int strip = strip0; strip <= strip1; strip++) {
+                final long rowsPerStripLong = 0xFFFFffffL & rowsPerStrip;
+                final long rowsRemaining = height - (strip * rowsPerStripLong);
+                final long rowsInThisStrip = Math.min(rowsRemaining, rowsPerStripLong);
+                final long bytesPerRow = (bitsPerPixel * width + 7) / 8;
+                final long bytesPerStrip = rowsInThisStrip * bytesPerRow;
+                final long pixelsPerStrip = rowsInThisStrip * width;
+
+                byte[] b = new byte[(int) bytesPerStrip];
+                for (int iPlane = 0; iPlane < 3; iPlane++) {
+                    int planeStrip = iPlane * nStripsInPlane + strip;
+                    final byte[] compressed = imageData.getImageData(planeStrip).getData();
+                    final byte[] decompressed = decompress(compressed, compression,
+                      (int) bytesPerStrip, width, (int) rowsInThisStrip);
+                    int index = iPlane;
+                    for (int i = 0; i < decompressed.length; i++) {
+                        b[index] = decompressed[i];
+                        index += 3;
+                    }
+                }
+                interpretStrip(workingBuilder, b, (int) pixelsPerStrip, height);
+            }
```

Comment:
This code here (`readImageData(final Rectangle subImage)`) looks a lot like the code above for `readImageData(final ImageBuilder imageBuilder)`. Should we make a function to read strips and re-use in both methods?

Reply:
These are two methods have a lot of redundancy and it would be better to combine them into one.  Unfortunately, they are both public APIs, so I'm not sure we could change their external appearance.   Of course, we could create a single, consolidated method that does the processing. Then we could rewrite the existing methods to be "wrapper" methods that called the new consolidated method.

I think it might be better to treat that change as a separate pull request.

I contributed the "sub-image" logic a number of years ago. At the time, there was a certain amount of resistance to the idea of a sub-image API, but I thought it was necessary because may GeoTIFFs are huge (I'd been working with some that are 20000-by-20000 pixels large).  Anyway, I think I left the original method in place because the code was simpler than the sub-image code and I thought it would serve as a reference for anyone who was trying to figure out how it worked.  But I think you are right and the time has come to clean things up.

Reply:
>I think it might be better to treat that change as a separate pull request.

:+1: